File: crawler/src/parser.ts
import * as cheerio from "cheerio";
import crypto from "crypto";

export interface ParseResult {
  title: string;
  description: string;
  snippet: string;
  language: string;
  contentHash: string;
  links: string[];
  sitemaps: string[];
}

export function normalizeUrl(url: string, baseUrl?: string): string {
  try {
    const u = new URL(url, baseUrl);
    u.hash = "";
    const params = [
      "utm_source",
      "utm_medium",
      "utm_campaign",
      "utm_term",
      "utm_content",
      "fbclid",
      "gclid",
    ];
    params.forEach((p) => u.searchParams.delete(p));

    u.searchParams.sort();

    let res = u.toString();
    return res;
  } catch (e) {
    return url;
  }
}

export function parseHtml(html: string, baseUrl: string): ParseResult {
  const $ = cheerio.load(html);

  $(
    "script, style, nav, footer, header, noscript, .cookie-banner, .ads",
  ).remove();

  const title =
    $('meta[property="og:title"]').attr("content") ||
    $("title").text() ||
    $("h1").first().text() ||
    "";

  const description =
    $('meta[name="description"]').attr("content") ||
    $('meta[property="og:description"]').attr("content") ||
    "";

  let snippet = "";
  $("p").each((_, el) => {
    const text = $(el).text().trim().replace(/\s+/g, " ");
    if (text.length > 40 && !snippet) {
      snippet = text;
    }
  });

  if (!snippet)
    snippet = $("body").text().trim().substring(0, 500).replace(/\s+/g, " ");

  let language = $("html").attr("lang")?.split("-")[0] || "en";
  if (language.length > 2) language = language.substring(0, 2);

  const cleanTitle = title.trim().substring(0, 300).replace(/\s+/g, " ");
  const cleanDesc = description.trim().substring(0, 300).replace(/\s+/g, " ");
  const cleanSnippet = snippet.trim().substring(0, 500);

  const hash = crypto
    .createHash("sha256")
    .update(html)
    .digest("hex")
    .substring(0, 16);

  const links: string[] = [];
  const sitemaps: string[] = [];

  $('link[rel="sitemap"]').each((_, el) => {
    const href = $(el).attr("href");
    if (href) sitemaps.push(normalizeUrl(href, baseUrl));
  });

  $("a[href]").each((_, el) => {
    let href = $(el).attr("href");
    if (!href) return;

    if (
      href.startsWith("#") ||
      href.startsWith("javascript:") ||
      href.startsWith("mailto:") ||
      href.startsWith("tel:")
    )
      return;

    const normalized = normalizeUrl(href, baseUrl);
    if (normalized.startsWith("http")) {
      if (normalized.endsWith(".xml") && normalized.includes("sitemap")) {
        sitemaps.push(normalized);
      } else {
        links.push(normalized);
      }
    }
  });

  return {
    title: cleanTitle,
    description: cleanDesc,
    snippet: cleanSnippet,
    language: language.toLowerCase(),
    contentHash: hash,
    links: [...new Set(links)],
    sitemaps: [...new Set(sitemaps)],
  };
}

export function parseSitemap(xml: string) {
  const $ = cheerio.load(xml, { xmlMode: true });
  const urls: string[] = [];
  const nestedSitemaps: string[] = [];

  $("url > loc").each((_, el) => {
    const url = $(el).text().trim();
    if (url) urls.push(url);
  });

  $("sitemap > loc").each((_, el) => {
    const url = $(el).text().trim();
    if (url) nestedSitemaps.push(url);
  });

  return {
    urls: [...new Set(urls)],
    nestedSitemaps: [...new Set(nestedSitemaps)],
  };
}

File: crawler/parser.go
package crawler

import (
	"crawler/config"
	"crypto/sha256"
	"encoding/xml"
	"fmt"
	"io"
	"net/url"
	"strings"

	"github.com/PuerkitoBio/goquery"
)

type SitemapURL struct {
	Loc     string `xml:"loc"`
	LastMod string `xml:"lastmod"`
}

type Sitemap struct {
	XMLName xml.Name     `xml:"urlset"`
	URLs    []SitemapURL `xml:"url"`
}

type SitemapIndex struct {
	XMLName  xml.Name     `xml:"sitemapindex"`
	Sitemaps []SitemapURL `xml:"sitemap"`
}

func ParseHTML(rawURL string, body io.Reader) (config.PageData, []string, []string, error) {
	doc, err := goquery.NewDocumentFromReader(body)
	if err != nil {
		return config.PageData{}, nil, nil, err
	}

	data := config.PageData{URL: rawURL}
	data.Title = strings.TrimSpace(doc.Find("title").Text())

	doc.Find("meta").Each(func(i int, s *goquery.Selection) {
		name, _ := s.Attr("name")
		property, _ := s.Attr("property")
		content, _ := s.Attr("content")

		if strings.ToLower(name) == "description" || strings.ToLower(property) == "og:description" {
			if data.Description == "" {
				data.Description = strings.TrimSpace(content)
			}
		}
	})

	doc.Find("script, style, nav, footer, header").Remove()
	bodyText := strings.TrimSpace(doc.Find("body").Text())
	bodyText = strings.Join(strings.Fields(bodyText), " ")
	if len(bodyText) > 250 {
		data.Snippet = bodyText[:247] + "..."
	} else {
		data.Snippet = bodyText
	}

	lang, exists := doc.Find("html").Attr("lang")
	if exists {
		data.Language = strings.ToLower(strings.Split(lang, "-")[0])
	} else {
		data.Language = "en"
	}

	nsfwTags := []string{"porn", "xxx", "sex", "adult", "nude"}
	fullText := strings.ToLower(data.Title + " " + data.Description)
	for _, tag := range nsfwTags {
		if strings.Contains(fullText, tag) {
			data.NSFW = true
			break
		}
	}

	var links []string
	var sitemaps []string
	baseURL, _ := url.Parse(rawURL)

	if baseHref, exists := doc.Find("base").Attr("href"); exists {
		if baseURLExtracted, err := url.Parse(baseHref); err == nil {
			baseURL = baseURL.ResolveReference(baseURLExtracted)
		}
	}

	doc.Find("a").Each(func(i int, s *goquery.Selection) {
		link, exists := s.Attr("href")
		if exists {
			resolved := resolveURL(baseURL, link)
			if resolved != "" {
				links = append(links, resolved)
			}
		}
	})

	doc.Find("link").Each(func(i int, s *goquery.Selection) {
		rel, _ := s.Attr("rel")
		if strings.ToLower(rel) == "sitemap" {
			href, exists := s.Attr("href")
			if exists {
				resolved := resolveURL(baseURL, href)
				if resolved != "" {
					sitemaps = append(sitemaps, resolved)
				}
			}
		}
	})

	data.ContentHash = generateHash(data)

	return data, links, sitemaps, nil
}

func resolveURL(base *url.URL, link string) string {
	u, err := url.Parse(link)
	if err != nil {
		return ""
	}

	resolved := base.ResolveReference(u)

	if resolved.Scheme != "http" && resolved.Scheme != "https" {
		return ""
	}

	resolved.Fragment = ""

	return resolved.String()
}

func generateHash(data config.PageData) string {
	h := sha256.New()
	h.Write([]byte(data.Title + data.Description))

	return fmt.Sprintf("%x", h.Sum(nil))
}

func ParseSitemap(body io.Reader) ([]SitemapURL, error) {
	data, err := io.ReadAll(body)
	if err != nil {
		return nil, err
	}

	var urlset Sitemap
	if err := xml.Unmarshal(data, &urlset); err == nil && len(urlset.URLs) > 0 {
		return urlset.URLs, nil
	}

	var index SitemapIndex
	if err := xml.Unmarshal(data, &index); err == nil && len(index.Sitemaps) > 0 {
		return index.Sitemaps, nil
	}

	return nil, fmt.Errorf("unknown sitemap format")
}
